### ISSUE TITLE
CPU target miscompiles scatter followed by copy on specific CPUs (?)

### ISSUE THREAD

**Issue by inducer** [pocl/pocl#979]
Running this code:

https://gist.github.com/inducer/d605a8f04196a4bcbf73835062f336ec

under specific circumstances (see below) prints
```
[-1 -1 -1 -1]
```
when it should print
```
[ 0  8 16 24]
```

I've so far found two sets of circumstances under which this happens:
- On Github CI, as documented in https://github.com/illinois-ceesd/mirgecom/pull/468#issuecomment-898806938, using pocl 1.7 `h11147fc_2` and LLVM 11.1.0 `hf817b99_2` from conda, as shown in this CI log: https://github.com/illinois-ceesd/mirgecom/pull/468/checks?check_run_id=3343181488.
- On my personal machine (Ice Lake client), right after clearing caches, using
    - pocl 1.6 (9b8e620) + LLVM 11 (Debian's LLVM 11.0.1-2)
    - pocl 1.7 (2359701) + LLVM 11 (Debian's LLVM 11.0.1-2)
    - pocl 1.7 (2359701) + LLVM 12 (Debian's LLVM 12.0.1-3)

  If run again, the program is pulled from PyOpenCL cache and created using `clCreateProgramWithBinary`, and the correct output is observed.

Other machines with broadly the same software as my personal machine but different processor architectures (Broadwell server, Haswell server, Ivy Bridge server) do not seem to suffer from this.

To clear caches, I use the following script:
```
rm -Rf ~/.pocl
rm -Rf ~/.cache/pocl
rm -Rf ~/.cache/pytools
rm -Rf ~/.cache/pyopencl/
rm -Rf ~/.nv/ComputeCache/
```

**Comment by inducer** [comment]
Additional context: The behavior disappears with
```
prg = cl.Program(ctx, SRC).build(options=['-cl-opt-disable'])
```
inserted in the relevant spot.

(Thanks, @kaushikcfd!)

**Comment by inducer** [comment]
FWIW, you can use [the tmate action](https://github.com/marketplace/actions/debugging-with-tmate) and [PyOpenCL's install instructions](https://documen.tician.de/pyopencl/misc.html#installing-pyopencl) to quickly create a reproducing environment.

**Comment by pjaaskel** [comment]
Interesting. Since it depends on the timing of the kernel execution (the binary loaded should be identical to what PoCL executes when going through the source), it sounds like a race condition at the host code side. My quick guess is that you read the output too early in your host thread since the build thread takes more time when not hitting the disk cache.

**Comment by inducer** [comment]
I sort of buy that, except inserting even a 10s sleep before the transfer back (longer than the whole execution of the program otherwise) does not change the outcome.

**Comment by pjaaskel** [comment]
Do you initialize the buffer to -1 by default? What if you disable the kernel cache, does it reproduce always? Does Valgrind yell?

**Comment by inducer** [comment]
> Do you initialize the buffer to -1 by default?

Yep, see [this line](https://gist.github.com/inducer/d605a8f04196a4bcbf73835062f336ec#file-scatter-then-copy-py-L28).

```
$ clcache; pycl pocl-icompose.py
[-1 -1 -1 -1]
$ clcache; PYOPENCL_NO_CACHE=1 pycl pocl-icompose.py
[ 0  8 16 24]
clcache; POCL_KERNEL_CACHE=1 pycl pocl-icompose.py
[-1 -1 -1 -1]
clcache; POCL_KERNEL_CACHE=1 PYOPENCL_NO_CACHE=1 pycl pocl-icompose.py
[ 0  8 16 24]
```
I don't quite understand why the executions with and without the PyOpenCL cache differ. On an empty cache, they do basically the same thing. The only tangible difference is that the caching one retrieves the binary.

**Comment by pjaaskel** [comment]
There might be a slight difference with client-side caching and PoCL-side caching on the end result: PoCL-side caching also stores and utilizes the specialized work-group functions (with whatever specializations have happened over execution) whereas if you load the binary earlier than executing the kernel (and thus do not invoke the specialization mechanism to produce the specialized versions over the different kernel command executions), it might end up having only the non-specialized (at least for the local size) one, that is, the "dynamic local size" version. I'm not sure is that currently then specialized any longer if starting from the binaries (even if it had the bitcode in the poclbin to do so), thus ending up sending a dynamic WG function to the vectorizer instead of the WG-size specialized one (with known WI loop bounds, which might help the vectorizer).

**Comment by franz** [comment]
There are probably two (unrelated?) bugs here. One is a bug in the pocl cache code, which resulted in dynamic wg binary being used (incorrectly) instead of static wg binary. It's fixed in my [release_1_8](https://github.com/franz/pocl/tree/release_1_8) branch.

The other bug is the miscompilation. Since the buffer is initialized to -1 and it prints -1, that might mean the kernel has no instructions, or is not run at all. This happened to me quite recently. A mismatch between calling convention of a caller and a callee (in a SPIR binary) resulted in LLVM  replacing the whole kernel by "llvm_unreachable" - but this was *only* happening on a particular Intel machine, my laptop had no problem with the same SPIR binary (both have same OS & exact same LLVM). You could check whether something similar is happening on that machine, by using `export POCL_LEAVE_KERNEL_COMPILER_TEMP_FILES=1` before running the test, finding a file named `parallel.bc` in pocl cache directory, and disassembling it with `llvm-dis parallel.bc`.

**Comment by franz** [comment]
@inducer OK, based on your comment i'm closing the issue.

The overhead of calling `clGetProgramInfo()`, with the CPU driver, is compilation (of LLVM IR to object files) of each kernel in the program (this is necessary because the binary must be usable in a PoCL without LLVM). This might be small compared to other build steps like linking / source-to-LLVM-IR which happen before.
